cpp/unreal_plugins/UrdfRobot/Source/UrdfRobot/UrdfRobotComponent.cpp

```diff
@@ -0,0 +1,311 @@
+{
+    std::map<std::string, ArrayDesc> observation_space;
+
+    if (Std::contains(observation_components_, "link_configurations")) {
+        for (auto link_component : LinkComponents) {
+            observation_space.merge(link_component->getObservationSpace());
+        }
+    }
+
+    if (Std::contains(observation_components_, "joint_configurations")) {
+        for (auto joint_component : JointComponents) {
+            observation_space.merge(joint_component->getObservationSpace());
+        }
+    }
+
+    return observation_space;
+}
+
+void UUrdfRobotComponent::applyAction(const std::map<std::string, std::vector<uint8_t>>& actions)
+{
+    if (Std::contains(action_components_, "control_joints")) {
+        std::map <std::string, std::vector<double>> actions_reinterpreted;
+        for (auto& action_component : actions) {
+            std::vector<double> action_component_data_reinterpreted = Std::reinterpretAs<double>(action_component.second);
+            actions_reinterpreted[action_component.first] = action_component_data_reinterpreted;
+        }
+        applyAction(actions_reinterpreted);
+    }
+}
+
+std::map<std::string, std::vector<uint8_t>> UUrdfRobotComponent::getObservation() const
+{
+    std::map<std::string, std::vector<uint8_t>> observation;
+
+    if (Std::contains(observation_components_, "link_configurations")) {
+        for (auto link_component : LinkComponents) {
+            observation.merge(link_component->getObservation());
+        }
+    }
+
+    if (Std::contains(observation_components_, "joint_configurations")) {
+        for (auto joint_component : JointComponents) {
+            observation.merge(joint_component->getObservation());
+        }
+    }
+
+    return observation;
+}
+
+void UUrdfRobotComponent::reset()
+{
+    for (auto link_component : LinkComponents) {
+        link_component->reset();
+    }
+}
+
+void UUrdfRobotComponent::initialize(const UrdfRobotDesc* robot_desc)
+{
+    bool promote_to_children = true;
+    for (auto link_component : LinkComponents) {
+        link_component->DestroyComponent(promote_to_children);
+    }
+
+    for (auto joint_component : JointComponents) {
+        joint_component->DestroyComponent(promote_to_children);
+    }
+
+    LinkComponents.Empty();
+    JointComponents.Empty();
+
+    SP_ASSERT(link_components_.empty());
+    SP_ASSERT(joint_components_.empty());
+
+    SP_ASSERT(robot_desc);
+
+    UrdfLinkDesc* root_link_desc = robot_desc->root_link_desc_;
+    SP_ASSERT(root_link_desc);
+
+    SP_ASSERT(!Std::containsSubstring(root_link_desc->name_, "."));
+    root_link_component_ = NewObject<UUrdfLinkComponent>(this, Unreal::toFName(root_link_desc->name_));
+    SP_ASSERT(root_link_component_);
+    root_link_component_->SetupAttachment(this);
+    root_link_component_->RegisterComponent();
+    // It is very important to setup attachments, and register components before we create more child components.
```

Comment:
Updated.

Reply:
Comment not needed because the thing it is warning about has been encapsulated into the helper function. Great example of self-documenting code being better than comments because it stays up-to-date automatically.